### PR TITLE
Implement new Component class and ObjectPool

### LINKED ISSUE #118: Make TagComponent inherit from Component
After I read PR #114, and I understand it correctly, `components` in `queries` can accept `TagComponent` not just `Component` and `NotComponent`. So type def should be:

```typescript
static queries: {
    [queryName: string]: {
      components: (Component | TagComponent | NotComponent)[],
      ...
    }
};
```

Why it's that. Because `TagComponent` doesn't inherited from `Component`. I think `TagComponent` should be the kind of `Component` that has only reset method.

### LINKED ISSUE #147: Why registerComponent?
Looking through the sources I wonder why this method exists?

It is not mentioned in the guides, only the API docs, and its use is not required, as a newly used Component gets auto-registered by the ComponentManager. 

But also inside the ComponentManager I see a bit of housekeeping going on:
https://github.com/MozillaReality/ecsy/blob/dev/src/ComponentManager.js#L18-L19

There is the `Components` hash, and the `numComponents` counter, but it seems both are unused outside of the class, and don't really service a meaningful purpose inside the class, AFAICT. (e.g. the counter is never read from)

So do I miss something, or could this be removed?

### PR DESCRIPTION
**Pull request by robertlong**
This PR is intended to make the component class more straightforward, predictable, and accessible to tooling.

Object pooling is now handled with `.copy()` and `.clone()` methods rather than `.reset()`. The Component class comes with a default implementation for the constructor, copy, and clone that is based on the component schema.

Components must now be registered, an error will be thrown if you attempt to add a component before it is registered. On component registration we validate the schema. Components must also now extend the `Component` class.

Schema types have a new interface. They must implement `name, default, copy, and clone`. Types are validated upon registration.

The `createComponentClass` helper has been removed. You should now use the `Component` class instead.

To Do:

- [ ] Update docs
  - [ ] Remove references to `createComponentClass()`
  - [ ] Update `createType()` docs
  - [ ] Update references to Component `.reset()`
  - [ ] Add new Component docs
  - [ ] Add new Entity `.copy()` and `.clone()` docs
  - [ ] Add `ObjectPool` docs
- [ ] Write more tests
  - [ ] Entity `.copy()`
  - [ ] Entity `.clone()`

Closes #147 
Closes #118

### CODE REVIEW

File: src/Entity.js [hunk 1]
```diff
@@ -114,21 +112,36 @@ export class Entity {
...
-  // Initialize the entity. To be used when returning an entity to the pool
   reset() {
-    this.id = nextId++;
-    this._world = null;
+    this.id = this._entityManager._nextEntityId++;
     this._ComponentTypes.length = 0;
     this.queries.length = 0;
-    this._components = {};
+
+    for (var componentName in this.components) {
```
Comment (fernandojsg) [inline src/Entity.js:139]: _components

**Comment by IfnotFr** [comment]
Great work with this ! Component schema definition with upcoming [proposal-class-fields](https://github.com/tc39/proposal-class-fields). Will make some really clear/consice Components :

```javascript
import { Component, Types } from "ecsy"

class MyComponent {
    static schema = {
        aNumber: { type: Types.Number },
        anArray: { type: Types.Array },
    }
}
```

**Comment by fernandojsg** [comment]
@IfnotFr indeed! currently it's a bit uglier but it's fine so far. but when the static will arrive both the Component definition and the System's queries will be prettier for sure :)

**Comment by fernandojsg** [comment]
I updated the issues regarding the documentation and tests that we should update
https://github.com/MozillaReality/ecsy/issues/190
https://github.com/MozillaReality/ecsy/issues/165

**Pull request merged**